onnxruntime/contrib_ops/webgpu/moe/hidden_state_gather.wgsl.template

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// local_idx is used to copy hidden state row, workgroup_idx is token_idx
+// workgroup_idx is the token index in this expert
+
+// in: hiddenstate_for_expert [num_experts, num_tokens]
+// in: hidden_state(vec4)
+// out: new_hidden_state(vec4) [used_by, hidden_size]
+// out: expert_tokens [used_by]
```

Comment:
An optimized idea (this can avoid the for loop in this shader): Dispatch this shader based on the total size of new_hidden_state by `.SetDispatchGroupSize((sizeOf(new_hidden_state) + workgroupsize - 1)/ workgroupsize)` instead of `.SetDispatchGroupSize(used_by)`. The shader code would be something like below:
```
let new_hidden_indices = new_hidden.offsetToIndices(global_idx);
let used_by_idx = new_hidden_indices[0];
let hidden_idx = new_hidden_indices[1];
let token_idx = hiddenstate_for_expert[expert_base + used_by_idx];
let src = hidden_state.getByIndices(vec2(token_idx, hidden_idx));
new_hidden.setByOffset(global_idx, src);
```
PS: It's ok to change it in follow up PRs.